include/ulight/impl/lang/xml.hpp

```diff
@@ -0,0 +1,51 @@
+#ifndef ULIGHT_XML_HPP
+#define ULIGHT_XML_HPP
+
+#include <cstddef>
+#include <functional>
+#include <set>
+#include <string_view>
+
+#include "html.hpp"
+
+namespace ulight::xml {
+
+/// @brief matches whitespace at the beginning of str and returns
+/// the length of the matched whitespace. If the start of str
+/// is not whitespace 0 is returned
+[[nodiscard]]
+std::size_t match_whitespace(std::u8string_view str);
+
+struct Name_Match_Result {
+
+    std::size_t length = 0;
+    std::set<std::size_t> error_indicies;
+};
+
+/// @brief matches a name at the beginning of str until the
+/// beginning of str satisfies the predicate is_stop_sequence.
+/// Characters that are not allowed in a name e.g & and > are
+/// marked for later highlighting.
+[[nodiscard]]
+Name_Match_Result match_name_permissive(
+    std::u8string_view str,
+    const std::function<bool(std::u8string_view)>& is_stop_sequence
```

Comment:
Using `std::function` for this purpose is pretty bad. If anything, this should be a function template taking some callable object, defined entirely in the source file.

The problem with `std::function` is that calling it can be pretty expensive, and creating it may involve dynamic allocations. It also requires `<functional>`, which is a pretty fat header.